addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -0,0 +1,424 @@
+        }
+        catch ( final IOException e )
+        {
+            logger.error( String.format( "[NPM] Json node parse failed for resource: %s. Reason: %s", resource,
+                                         e.getMessage() ), e );
+            response = formatResponse( e, builderModifier );
+
+            return;
+        }
+        try
+        {
+            if ( target1 == null || target2 == null )
+            {
+                return;
+            }
+            out1 = target1.openOutputStream( TransferOperation.UPLOAD, true, eventMetadata );
+            out1.write( content1.getBytes() );
+
+            out2 = target2.openOutputStream( TransferOperation.UPLOAD, true, eventMetadata );
+            out2.write( Base64.decodeBase64( content2 ) );
+
+            setGeneratedTransfers( transfer, target1, target2 );
+        }
+        catch ( final IOException e )
+        {
+            logger.error( String.format( "[NPM] Failed to store the generated targets: s% and s%. Reason: s%",
+                                         target1.getResource(), target2.getResource(), e.getMessage() ), e );
+            response = formatResponse( e, builderModifier );
+        }
+        finally
+        {
+            closeQuietly( out1 );
+            closeQuietly( out2 );
+        }
+    }
+
+    public void generateHttpMetadataHeaders( final Transfer transfer, final HttpServletRequest request,
+                                             final Response response )
+    {
+        if ( transfer == null || request == null || response == null )
+        {
+            return;
+        }
+
+        Response responseWithLastModified =
+                response.fromResponse( response ).lastModified( new Date( transfer.lastModified() ) ).build();
+
+        Transfer metaTxfr = transfer.getSiblingMeta( HttpExchangeMetadata.FILE_EXTENSION );
+        if ( metaTxfr == null )
+        {
+            if ( transfer.isDirectory() )
+            {
+                metaTxfr = transfer.getChild( HttpExchangeMetadata.FILE_EXTENSION );
+            }
+            else
+            {
+                return;
+            }
+        }
+
+        final HttpExchangeMetadata metadata = new StoreHttpExchangeMetadata( request, responseWithLastModified );
+        OutputStream out = null;
+        try
+        {
+            out = metaTxfr.openOutputStream( TransferOperation.GENERATE, false );
+            out.write( mapper.writeValueAsBytes( metadata ) );
+        }
+        catch ( final IOException e )
+        {
+            logger.error( "Failed to write metadata for HTTP exchange to: {}. Reason: {}", metaTxfr, e );
+        }
+        finally
+        {
+            IOUtils.closeQuietly( out );
+        }
+
+        // npm will generate .tgz and version json metadata files from the package json file target,
+        // which will also need the HttpExchangeMetadata for npm header check.
+
+        List<Transfer> generated = getGeneratedTransfers( transfer );
```

Comment:
Use a method param to pass these in, please. Keeping the state scoped as locally as possible is critical for memory performance.